src/inmanta/resources.py

```diff
@@ -301,10 +407,17 @@ def map_field(
                 value = cls.map[field_name](exporter, model_object)
             else:
                 value = getattr(model_object, field_name)
-            # serialize to weed out all unknowns
-            # not very efficient, but the tree has to be traversed anyways
-            # passing along the serialized version would break the resource apis
-            json.dumps(value, default=inmanta.util.api_boundary_json_encoder)
+
+            # walk the entire model to find any value references. Additionally we also want to make sure we raise exceptions on:
+            # - Unknowns
+            # - DynamicProxys to entities
+            if reference_collector:
+                value = collect_references(reference_collector, value, field_name)
+            else:
+                # use json dump when there is no reference_collector
+                # TODO: can't we just always use collect_references?
```

Comment:
Handle this TODO. We could adapt collect_references to also work with None as reference_collector and only the serialization checks.

Alternatively we remove the TODO

Reply:
nice